File: android/app/src/main/java/com/risa/app/NativeCrashHandler.java
package com.risa.app;

import android.util.Log;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

/**
 * Minimal native crash logger to capture uncaught exceptions before the process dies.
 * Writes to external app storage: /sdcard/Android/data/com.risa.app/files/logs/native_crash.txt
 */
public class NativeCrashHandler implements Thread.UncaughtExceptionHandler {
    private static final String TAG = "NativeCrashHandler";
    private final Thread.UncaughtExceptionHandler defaultHandler;
    private final File logFile;

    public NativeCrashHandler(File externalFilesDir, Thread.UncaughtExceptionHandler defaultHandler) {
        this.defaultHandler = defaultHandler;
        File logDir = new File(externalFilesDir, "logs");
        if (!logDir.exists()) {
            //noinspection ResultOfMethodCallIgnored
            logDir.mkdirs();
        }
        this.logFile = new File(logDir, "native_crash.txt");
    }

    @Override
    public void uncaughtException(Thread t, Throwable e) {
        writeCrash(e);
        if (defaultHandler != null) {
            defaultHandler.uncaughtException(t, e);
        }
    }

    private void writeCrash(Throwable e) {
        FileWriter writer = null;
        try {
            writer = new FileWriter(logFile, true);
            writer.write("\n--- " + System.currentTimeMillis() + " [native] ---\n");
            writer.write(Log.getStackTraceString(e));
            writer.write("\n");
            writer.flush();
        } catch (IOException ignored) {
        } finally {
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException ignored) {
                }
            }
        }
    }
}

File: android/app/src/main/kotlin/com/risa/app/NativeCrashHandler.kt
package com.risa.app

import android.util.Log
import java.io.File
import java.io.PrintWriter
import java.io.StringWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class NativeCrashHandler(
    private val externalDir: File,
    private val defaultHandler: Thread.UncaughtExceptionHandler?
) : Thread.UncaughtExceptionHandler {

    override fun uncaughtException(thread: Thread, throwable: Throwable) {
        try {
            writeCrashToFile(thread, throwable)
        } catch (e: Exception) {
            Log.e("NativeCrashHandler", "Failed to write crash log", e)
        }
        
        // Call the default handler to ensure proper crash handling
        defaultHandler?.uncaughtException(thread, throwable)
    }

    private fun writeCrashToFile(thread: Thread, throwable: Throwable) {
        val timestamp = System.currentTimeMillis()
        val logsDir = File(externalDir, "logs")
        if (!logsDir.exists()) {
            logsDir.mkdirs()
        }
        
        val crashFile = File(logsDir, "native_crash.txt")
        
        try {
            val stackTrace = StringWriter()
            throwable.printStackTrace(PrintWriter(stackTrace))
            
            val crashInfo = """
--- $timestamp [native] ---
${stackTrace}
"""
            
            crashFile.writeText(crashInfo)
            Log.i("NativeCrashHandler", "Crash written to: ${crashFile.absolutePath}")
        } catch (e: Exception) {
            Log.e("NativeCrashHandler", "Failed to write crash file", e)
        }
    }
}